src/web.rs

```diff
@@ -0,0 +1,94 @@
+use log::{debug, info};
+use std::convert::TryInto;
+use std::sync::{Arc, RwLock};
+
+use hyper::service::{make_service_fn, service_fn};
+use hyper::{Body, Method, Request, Response, Server, StatusCode};
+
+use anyhow::{anyhow, Context, Error};
+
+use crate::expectation::model::{Expectation, HttpMethod};
+use crate::mock::{look_for_expectation, IncomingRequest};
+
+impl TryInto<HttpMethod> for &Method {
+    type Error = Error;
+
+    fn try_into(self) -> Result<HttpMethod, Self::Error> {
+        match self {
+            &Method::GET => Ok(HttpMethod::GET),
+            &Method::POST => Ok(HttpMethod::POST),
+            &Method::PUT => Ok(HttpMethod::PUT),
+            &Method::DELETE => Ok(HttpMethod::DELETE),
+            &Method::HEAD => Ok(HttpMethod::HEAD),
+            method => Err(anyhow!("{} isn't managed as HttpMethod", method)),
+        }
+    }
+}
+
+impl TryInto<Response<Body>> for &Expectation {
+    type Error = anyhow::Error;
+
+    fn try_into(self) -> Result<Response<Body>, Error> {
+        Response::builder()
+            .status(self.response.status_code.unwrap_or(200))
+            .body(
+                self.response
+                    .body
+                    .as_ref()
+                    .map(|body| Body::from(body.clone()))
+                    .unwrap_or(Body::empty()),
+            )
+            .context(format!("Error creating http response for {:?}", self))
+    }
+}
+
+pub struct State {
+    pub expectations: Vec<Expectation>,
+}
+
+async fn handler<T>(req: Request<T>, state: Arc<RwLock<State>>) -> Result<Response<Body>, Error> {
+    let incoming_request = req.method().try_into().map(|method| IncomingRequest {
```

Comment:
I don't know what's the state of the art of error handling in rust but IMO `try_into` isn't a really clear way to perform.

Reply:
Going to change around `TryFrom` but we have a limitation to go for a `Into` or a `From`.

If we implement `impl From<Method> for Option<HttpMethod>` as `Method` and `Option` aren't from this crate the compilation will fail so we have to manage the fact that HttpMethod doesn't modelize all ths hyper avaiable in method.